internal/types/subtypes/interceptor.go

```diff
@@ -0,0 +1,193 @@
+package subtypes
+
+import (
+	"context"
+
+	"github.com/hashicorp/boundary/sdk/pbs/controller/protooptions"
+	"google.golang.org/grpc"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/types/descriptorpb"
+)
+
+func messageDomain(m proto.Message) string {
+	r := m.ProtoReflect()
+	fd := r.Descriptor().ParentFile()
+	if fd == nil {
+		return ""
+	}
+
+	opts, ok := fd.Options().(*descriptorpb.FileOptions)
+	if !ok {
+		return ""
+	}
+
+	domain := proto.GetExtension(opts, protooptions.E_Domain).(string)
+	return domain
+}
+
+func transformRequestAttributes(_ context.Context, req proto.Message) proto.Message {
+	domain := messageDomain(req)
+
+	r := req.ProtoReflect()
+	fields := r.Descriptor().Fields()
+
+	itemField := fields.ByName("item")
+	idField := fields.ByName("id")
+	attributesField := fields.ByName("attributes")
+
+	fieldValue := func(m protoreflect.Message, fd protoreflect.FieldDescriptor) string {
+		if fd == nil {
+			return ""
+		}
+		return m.Get(fd).String()
+	}
+
+	var st Subtype
+	switch {
+	case itemField != nil:
+		itemR := itemField.Message()
+		if itemR == nil {
+			return req
+		}
+		id := fieldValue(r, idField)
+
+		item := r.Get(itemField).Message().Interface()
+		itemFields := itemR.Fields()
+
+		typeField := itemFields.ByName("type")
+		t := fieldValue(item.ProtoReflect(), typeField)
+
+		sourceIdField := sourceIdFieldDescriptor(item.ProtoReflect().Descriptor())
+		sourceId := fieldValue(item.ProtoReflect(), sourceIdField)
+
+		switch {
+		case idField != nil && id != "":
+			st = SubtypeFromId(domain, id)
+		case sourceIdField != nil && sourceId != "":
+			st = SubtypeFromId(domain, sourceId)
+		case typeField != nil && t != "":
+			st = Subtype(t)
+		default: // need either type or id
+			return req
+		}
+		if err := convertAttributesToSubtype(item, st); err != nil {
+			return req
+		}
+	case idField != nil && attributesField != nil:
+		id := r.Get(idField).String()
+		st = SubtypeFromId(domain, id)
+		if err := convertAttributesToSubtype(req, st); err != nil {
+			return req
+		}
+	}
+	return req
+}
+
+func transformResponseItemAttributes(item proto.Message) error {
+	r := item.ProtoReflect()
+
+	itemFields := r.Descriptor().Fields()
+	typeField := itemFields.ByName("type")
+	if typeField == nil {
+		// not an item with subtypes
+		return nil
+	}
+
+	st := Subtype(item.ProtoReflect().Get(typeField).String())
+	return convertAttributesToDefault(item, st)
+}
+
+func transformResponseAttributes(_ context.Context, res proto.Message) proto.Message {
+	r := res.ProtoReflect()
+	fields := r.Descriptor().Fields()
+
+	itemField := fields.ByName("item")
+	itemsField := fields.ByName("items")
```

Comment:
Could we add a comment for the scenarios we expect here similarly to the request?

Reply:
See [7658e9e](https://github.com/hashicorp/boundary/pull/1967/commits/7658e9e489454006241899d5864127cb99e47422)